### PR TITLE
Speed up database operations

### PR DESCRIPTION
**Pull request by HakierGrzonzo**
# What was done:

- Insert operations where _clumped together_ instead of performing inserts per each row. In order to handle big operations efficiently, `LOAD DATA LOCAL INFILE` was used.
 -  local_infile is automatically set if it was disabled.
- Stories are converted in 20 parallel threads and database connections

# What was achieved 

Around 7 times improvement in processing speed:
```
6:26.72 - with patches
85.56user 4.02system 6:26.72elapsed 23%CPU (0avgtext+0avgdata 45324maxresident)k
72inputs+165592outputs (0major+53482minor)pagefaults 0swaps


47:09.18 - without patch (current master)
98.07user 3.07system 47:09.18elapsed 3%CPU (0avgtext+0avgdata 206220maxresident)k
0inputs+168104outputs (0major+116977minor)pagefaults 0swaps
```

### CODE REVIEW

File: efiction/metadata.py [hunk 1]
```diff
@@ -71,17 +71,29 @@ def _convert_authors(self, old_authors=None):
         self.logger.info("Converting authors...")
         current = 0
         total = len(old_authors)
+        query = f"INSERT INTO authors (`id`, `name`, `email`) Values " 
+        query_lines = []
         for old_author in old_authors:
```
Comment (passeriforma) [inline efiction/metadata.py:82]: Could add another comment here explaining what's happening in the loop? (given your commenting elsewhere is excellent)

File: efiction/metadata.py [hunk 2]
```diff
@@ -137,26 +138,35 @@ def _convert_story_tags(self, old_story):
...
-    def _convert_tags_join(self, new_story, tags):
+    def _convert_tags_join(self, new_story, tags, sql=None):
         full_query = f"INSERT INTO item_tags (item_id, item_type, tag_id) VALUES "
         tag_query = []
         for tag_list in tags.values():
             for tag in tag_list:
                 tag_query.append(f"""{new_story['id'], "story", tag}""")
         full_query += ", ".join(tag_query)
-        self.sql.execute(self.working_open_doors, full_query)
+        if sql is None:
```
Comment (passerinea) [inline efiction/metadata.py]: Given you repeat these four lines three times that I can see in here, consider making it a separate function to help prevent human error going forwards.

File: efiction/metadata.py [hunk 3]
```diff
@@ -189,20 +203,24 @@ def convert_stories(self, language_code):
...
             self.logger.debug(f"  authors...")
-            self._convert_author_join(new_story, old_story['uid'])
+            self._convert_author_join(new_story, old_story['uid'], sql)
             # Find if there are any coauthors for the work
-            coauthors = self.fetch_coauthors(new_story)
+            coauthors = self.fetch_coauthors(new_story, sql)
             for coauthor in coauthors:
-                self._convert_author_join(new_story, coauthor)
+                self._convert_author_join(new_story, coauthor, sql)
+            #current = print_progress(current, total, "stories converted")
```
Comment (passerinea) [inline efiction/metadata.py]: Is this meant to be commented out?

File: efiction/metadata.py (hunk 1 above)
Comment (passerinea) [inline efiction/metadata.py:82]: Could add another comment here explaining what's happening in the loop? (given your commenting elsewhere is excellent)

File: efiction/chapters.py [hunk 4]
```diff
@@ -52,38 +52,55 @@ def __load_chapter_text_into_db(self, chapter_paths: List[dict]):
...
+                    #                  old_chapter['sid'], old_chapter['notes']))
+                    query_file.write("{}\t{}\t{}\t{}\t{}\t{}\n".format(
+                        chapid,
+                        old_chapter['inorder'],
+                        old_chapter['title'],
+                        text.replace("\t", " " * 4),
+                        old_chapter['sid'],
+                        old_chapter['notes']
+                    ))
+                current = print_progress(current, total, "chapters converted")
```
Comment (passerinea) [inline efiction/chapters.py]: Should this be here, or after the query has been executed? (i.e. on line 104)

File: opendoors/big_insert.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+from .mysql import SqlDb
+from tempfile import NamedTemporaryFile
+
+class BigInsert:
+    def __init__(self, database: str, table_name: str, columns: list, sql: SqlDb):
+        """
```
Comment (passerinea) [inline opendoors/big_insert.py:6]: Beautiful commenting

File: efiction/chapters.py [hunk 6]
```diff
@@ -52,56 +53,53 @@ def __load_chapter_text_into_db(self, chapter_paths: List[dict]):
...
-            INTO TABLE chapters
-            fields terminated by '\t'
-            lines terminated by '\n'
-            (id, position, title, text, story_id, notes);
-        """)
+                text = normalize(raw)
+                if key_find('endnotes', old_chapter):
+                    text = text + f"\n\n\n<hr>\n{old_chapter['endnotes']}"
 
+                #query = """
```
Comment (passerinea) [inline efiction/chapters.py]: Remove commented out code

**Review by passerinea — APPROVED** [review]
The logic looks solid (and it's going to be so exciting to have multithreading for this!) - added a few comments, but nothing major.

LGTM

File: efiction/metadata.py (hunk 2 above)
Comment (HakierGrzonzo) [reply efiction/metadata.py]: I could make it a decorator, but I don't think the added complexity would be worth it.
For now I will change it to something like this:

```python
if sql is None:
    sql = self.sql
```

And comment it appropriately

**Review by ariana-paris — APPROVED** [review]

**Pull request merged**